replay/models/nmslib_hnsw.py

```diff
@@ -0,0 +1,646 @@
+            if index_type == "sparse":
+                item_vectors = item_vectors.toPandas()
+
+                index = nmslib.init(  # pylint: disable=c-extension-no-member
+                    method=params["method"],
+                    space=params["space"],
+                    data_type=nmslib.DataType.SPARSE_VECTOR,  # pylint: disable=c-extension-no-member
+                )
+
+                data = item_vectors["similarity"].values
+                row_ind = item_vectors["item_idx_two"].values
+                col_ind = item_vectors["item_idx_one"].values
+
+                sim_matrix = csr_matrix(
+                    (data, (row_ind, col_ind)),
+                    shape=(items_count, items_count),
+                )
+                index.addDataPointBatch(data=sim_matrix)
+                index.createIndex(index_params)
+
+                # saving index to local temp file and sending it to executors
+                temp_dir = tempfile.mkdtemp()
+                weakref.finalize(self, shutil.rmtree, temp_dir)
+                self.__dict__.pop('_spark_index_file_uid', None)
+                tmp_file_path = os.path.join(
+                    temp_dir, f"{INDEX_FILENAME}_{self._spark_index_file_uid}"
+                )
+                index.saveIndex(tmp_file_path, save_data=True)
+                spark = SparkSession.getActiveSession()
+                # for the "sparse" type we need to store two files
+                spark.sparkContext.addFile("file://" + tmp_file_path)
+                spark.sparkContext.addFile("file://" + tmp_file_path + ".dat")
+
+            else:
+                item_vectors = item_vectors.toPandas()
+                item_vectors_np = np.squeeze(item_vectors[features_col].values)
+                index = nmslib.init(  # pylint: disable=c-extension-no-member
+                    method=params["method"],
+                    space=params["space"],
+                    data_type=nmslib.DataType.DENSE_VECTOR,  # pylint: disable=c-extension-no-member
+                )
+                index.addDataPointBatch(
+                    data=np.stack(item_vectors_np),
+                    ids=item_vectors["item_idx"].values,
+                )
+                index.createIndex(index_params)
+
+                # saving index to local temp file and sending it to executors
+                temp_dir = tempfile.mkdtemp()
+                weakref.finalize(self, shutil.rmtree, temp_dir)
+                tmp_file_path = os.path.join(
+                    temp_dir, f"{INDEX_FILENAME}_{self._spark_index_file_uid}"
+                )
+                index.saveIndex(tmp_file_path)
+                spark = SparkSession.getActiveSession()
+                spark.sparkContext.addFile("file://" + tmp_file_path)
+
+    def _update_hnsw_index(
+        self,
+        item_vectors: DataFrame,
+        features_col: str,
+        params: Dict[str, Any],
+    ):
+        index = nmslib.init(    # pylint: disable=c-extension-no-member
```

Comment:
please think of single responsibility principle, here it means there multiple code entities that can build index

Reply:
Thanks, I'll try to make it into separate classes.